state/watcher.go

```diff
@@ -803,19 +803,37 @@ func (w *minUnitsWatcher) Changes() <-chan []string {
 // WatchModelMachinesCharmProfiles returns a StringsWatcher that notifies of
 // changes to the upgrade charm profile charm url for a machine.
 func (st *State) WatchModelMachinesCharmProfiles() (StringsWatcher, error) {
-	isMachineRegexp := fmt.Sprintf("^%s:%s$", st.ModelUUID(), names.NumberSnippet)
-	return st.watchCharmProfiles(isMachineRegexp)
+	isMachineRegexp := fmt.Sprintf("^%s:%s#%s$", st.ModelUUID(), names.NumberSnippet, names.UnitSnippet)
+	return st.watchCharmProfiles(isMachineRegexp, true)
 }
 
 // WatchContainersCharmProfiles starts a StringsWatcher to notify when
 // the provisioner should update the charm profiles used by any container on
 // the machine.
 func (m *Machine) WatchContainersCharmProfiles(ctype instance.ContainerType) (StringsWatcher, error) {
-	isChildRegexp := fmt.Sprintf("^%s/%s/%s$", m.doc.DocID, ctype, names.NumberSnippet)
-	return m.st.watchCharmProfiles(isChildRegexp)
+	isChildRegexp := fmt.Sprintf("^%s/%s/%s#%s$", m.doc.DocID, ctype, names.NumberSnippet, names.UnitSnippet)
+	return m.st.watchCharmProfiles(isChildRegexp, true)
 }
 
-func (st *State) watchCharmProfiles(regExp string) (StringsWatcher, error) {
+// WatchModelMachinesCharmProfiles returns a StringsWatcher that notifies of
+// changes to the upgrade charm profile charm url for a machine.
+//
+// TODO - is this needed?  2019-02-01
```

Comment:
Is this `TODO` relevant...

Reply:
I've resolved the TODO, removing WatchModelMachinesCharmProfiles as it's no longer called.  Only WatchContainersCharmProfiles is needed for compatibility.